fw/cache.c

```diff
@@ -2838,14 +2861,48 @@ tfw_cache_build_resp(TfwHttpReq *req, TfwCacheEntry *ce, long lifetime)
 	if (tfw_cache_set_status(db, ce, resp, &trec, &p, &h_len))
 		goto free;
 
-	for (h = TFW_HTTP_HDR_REGULAR; h < ce->hdr_num; ++h) {
-		bool skip = !TFW_MSG_H2(req) && (h >= ce->hdr_h2_off);
+	trailer_off = TFW_MSG_H2(req) ? ce->trailer_off : ce->hdr_num;
+	for (h = TFW_HTTP_HDR_REGULAR; h < trailer_off; ++h) {
+		bool skip = !TFW_MSG_H2(req) && (h >= ce->hdr_h2_off)
+			     && (h < ce->trailer_off);
 
 		if (tfw_cache_build_resp_hdr(db, resp, h_mods, &trec, &p,
 					     &h_len, skip))
 			goto free;
 	}
 
+	if (unlikely(trailer_off < ce->hdr_num)) {
+		TfwHttpTransIter pmit, nmit;
+		struct sk_buff *pskb_head = resp->msg.skb_head;
+
+		has_trailers = true;
+		pmit = nmit = resp->mit;
+		nmit.start_off = 0;
+		nmit.iter.skb = nmit.iter.skb_head = NULL;
+		nmit.iter.frag = -1;
+
+		resp->msg.skb_head = NULL;
```

Comment:
I don't agree with this solution. We should rework our standard functions to add trailers. Here we have extra loop over skbs and it looks bad for me. @krizhanovsky @const-t please look he and write your opinion.

Reply:
Can't find a way to eliminate this `skb_head` operation, because the response built from cache requires the headers to be built first, so there's no way to avoid creating a new skb before building the body and appending it after the body.

Reply:
Remove the skb_head manipulation loop: [commit](https://github.com/tempesta-tech/tempesta/pull/2173/commits/6bcd9e49065c028403a1c93b5722e3cbaea366d3).

Reply:
I try to rewrite cache functions to make them able to work with trailers in the proper way next Friday (review day).

Reply:
Ok, looking forward to your suggested commits which I can then push directly, thanks.

Reply:
@EvgeniiMekhanik Done. https://github.com/tempesta-tech/tempesta/pull/2173/commits/4932ecbd306c5cdc3e80d8198ab6d2de869a9389